benchmarks/advection_reaction_3D/kokkos/ParallelGrid.hpp

```diff
@@ -0,0 +1,596 @@
+      int nbcoords[] = {coords[0]+1, coords[1], coords[2]};
+      int retval = MPI_Cart_rank(cart_comm, nbcoords, &ipE);
+      assert(retval == MPI_SUCCESS);
+    }
+
+    /* Allocate send/receive buffers and determine ID for communication South */
+    if (upwindRight) {
+      Srecv_  = Vec1D("Srecv", dof*width*nxl*nzl);
+      SrecvH_ = Kokkos::create_mirror_view(Srecv_);
+    } else {
+      Ssend_  = Vec1D("Ssend", dof*width*nxl*nzl);
+      SsendH_ = Kokkos::create_mirror_view(Ssend_);
+    }
+    ipS = MPI_PROC_NULL;
+    if ((coords[1] > 0) || (bc == BoundaryType::PERIODIC)) {
+      int nbcoords[] = {coords[0], coords[1]-1, coords[2]};
+      int retval = MPI_Cart_rank(cart_comm, nbcoords, &ipS);
+      assert(retval == MPI_SUCCESS);
+    }
+
+    /* Allocate send/receive buffers and determine ID for communication North */
+    if (upwindRight) {
+      Nsend_  = Vec1D("Nsend", dof*width*nxl*nzl);
+      NsendH_ = Kokkos::create_mirror_view(Nsend_);
+    } else {
+      Nrecv_  = Vec1D("Nrecv", dof*width*nxl*nzl);
+      NrecvH_ = Kokkos::create_mirror_view(Nrecv_);
+    }
+    ipN = MPI_PROC_NULL;
+    if ((coords[1] < dims[1]-1) || (bc == BoundaryType::PERIODIC)) {
+      int nbcoords[] = {coords[0], coords[1]+1, coords[2]};
+      int retval = MPI_Cart_rank(cart_comm, nbcoords, &ipN);
+      assert(retval == MPI_SUCCESS);
+    }
+
+    /* Allocate send/receive buffers and determine ID for communication Back */
+    if (upwindRight) {
+      Brecv_  = Vec1D("Brecv", dof*width*nxl*nyl);
+      BrecvH_ = Kokkos::create_mirror_view(Brecv_);
+    } else {
+      Bsend_  = Vec1D("Bsend", dof*width*nxl*nyl);
+      BsendH_ = Kokkos::create_mirror_view(Bsend_);
+    }
+    ipB = MPI_PROC_NULL;
+    if ((coords[2] > 0) || (bc == BoundaryType::PERIODIC)) {
+      int nbcoords[] = {coords[0], coords[1], coords[2]-1};
+      int retval = MPI_Cart_rank(cart_comm, nbcoords, &ipB);
+      assert(retval == MPI_SUCCESS);
+    }
+
+    /* Allocate send/receive buffers and determine ID for communication Front */
+    if (upwindRight) {
+      Fsend_  = Vec1D("Fsend", dof*width*nxl*nyl);
+      FsendH_ = Kokkos::create_mirror_view(Fsend_);
+    } else {
+      Frecv_  = Vec1D("Frecv", dof*width*nxl*nyl);
+      FrecvH_ = Kokkos::create_mirror_view(Frecv_);
+    }
+    ipF = MPI_PROC_NULL;
+    if ((coords[2] < dims[2]-1) || (bc == BoundaryType::PERIODIC)) {
+      int nbcoords[] = {coords[0], coords[1], coords[2]+1};
+      int retval = MPI_Cart_rank(cart_comm, nbcoords, &ipF);
+      assert(retval == MPI_SUCCESS);
+    }
+
+  }
+
+  // Initiate non-blocking neighbor communication
+  int ExchangeStart()
+  {
+    int retval = 0;
+    nreq = 0;
+
+    // Initialize all requests in array
+    for (int i=0; i<12; i++)
+      req[i] = MPI_REQUEST_NULL;
+
+    // Open an Irecv buffer on host for each neighbor
+    if ((ipW != MPI_PROC_NULL) && (upwindRight))
+    {
+      retval = MPI_Irecv(WrecvH_.data(), dof*width*nyl*nzl, MPI_SUNREALTYPE, ipW,
```

Comment:
I left `width` in here as a placeholder in case we decide to implement higher-order advection stencils (this was one of Cody's "to-do" items).  That said, supporting a higher-order advection stencil would require so many other changes that leaving just this piece in seems unnecessary.  I'll go ahead and remove.